### PR TITLE
migration: Add a case about ovs bridge

### PR DESCRIPTION
**Pull request by Yingshun**
This PR adds a case -- migrate guest with bridge type interface
connected to ovs bridge.

depends on https://github.com/avocado-framework/avocado-vt/pull/2738
https://github.com/avocado-framework/avocado-vt/pull/2703
Signed-off-by: Yingshun Cui <yicui@redhat.com>

### CODE REVIEW

**Comment by Yingshun** [comment]
VM with special network settings always failed to get ip address with get_address(), so I create a new .py for network related migration tests.
```
# avocado run --vt-type libvirt --vt-machine-type q35 virsh.migrate_network.ovs_br.without_postcopy 
JOB ID     : 62d468a6b8cec8df76e92c270faff53db66efb68
JOB LOG    : /root/avocado/job-results/job-2020-08-12T01.31-62d468a/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virsh.migrate_network.ovs_br.without_postcopy: PASS (316.58 s)
RESULTS    : PASS 1 | 
```

**Comment by Yingshun** [comment]
@dzhengfy It's a high priority case.

File: libvirt/tests/src/migration/migrate_network.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+def run(test, params, env):
+    """
+    Test migration with special network settings
+    1) migrate guest with bridge type interface connected to ovs bridge
+
+    :param test: test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    def check_vm_network_accessed(ping_dest, session=None):
```
Comment (dzhengfy) [inline libvirt/tests/src/migration/migrate_network.py:33]: If you did not find a share library for pinging the localhost/local vm, it is time to create it. 
I recommend we add the new function in virttest/utils_libvirt/libvirt_network.py

BTW,  please replace 'utils_test.ping' with 'utils_net.ping'. Functions in utils_test.py may be removed in some day.

File: libvirt/tests/src/migration/migrate_network.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+        logging.info("Check VM network connectivity")
+        status, output = utils_test.ping(ping_dest,
+                                         count=10,
+                                         timeout=20,
+                                         output_func=logging.debug,
+                                         session=session)
+        if status != 0:
+            test.fail("Ping failed, status: %s, output: %s" % (status, output))
+
+    def update_iface_xml(vm_name, iface_dict):
```
Comment (dzhengfy) [inline libvirt/tests/src/migration/migrate_network.py:52]: Check libvirt.modify_vm_iface() to see if it meets your needs.

File: libvirt/tests/src/migration/migrate_network.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+            runner = session.cmd
+        iface_name = utils_net.get_net_if(runner=runner,
+                                          state="UP")[0]
+        if not utils_package.package_install(["tmux", "openvswitch2.11"], session):
+            test.error("Failed to install the required packages.")
+
+        runner = process.run
+        if session:
+            runner = remote.RemoteRunner(session=session).run
+        service.Factory.create_service("openvswitch", runner).restart()
```
Comment (dzhengfy) [inline libvirt/tests/src/migration/migrate_network.py:92]: There is openswitch.py which has many utilities for ovs. FYI.
If it is hard for you to reuse above classes, I think at least we should put this function in virttest/utils_net.py.

File: libvirt/tests/src/migration/migrate_network.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+            else:
+                mac = iface_dict["mac"]
+
+            update_iface_xml(vm_name, iface_dict)
+
+        # Change the disk of the vm
+        libvirt.set_vm_disk(vm, params)
+
+        if not vm.is_alive():
+            vm.start()
```
Comment (dzhengfy) [inline libvirt/tests/src/migration/migrate_network.py:205]: Use try..except..finally to handle unexpect vm start failure.

File: libvirt/tests/src/migration/migrate_network.py (hunk 2 above)
Comment (Yingshun) [reply libvirt/tests/src/migration/migrate_network.py:52]: It used libvirt.modify_vm_iface() in L65.

File: libvirt/tests/src/migration/migrate_network.py (hunk 3 above)
Comment (Yingshun) [reply libvirt/tests/src/migration/migrate_network.py]: Updated.

File: libvirt/tests/src/migration/migrate_network.py (hunk 1 above)
Comment (Yingshun) [reply libvirt/tests/src/migration/migrate_network.py:34]: Thanks for your reminder, updated to utils_net.ping. But I wonder if it's worth to add a new function into avocado-vt? It just calls utils_net.ping and check the result if it needs to raise, it's really a simple function I think. Do you mind sharing your thoughts?

File: libvirt/tests/src/migration/migrate_network.py (hunk 2 above)
Comment (Yingshun) [reply libvirt/tests/src/migration/migrate_network.py:47]: It's L60 for now.

Comment (dzhengfy) [reply libvirt/tests/src/migration/migrate_network.py:47]: Yes. My understanding is that you want to remove all existing nics and add a new nic. 
From my POV, modify_vm_iface() is messed up a little now. I am thinking if we need create a new function for creating a nic and reuse the assembling dict in modify_vm_iface(). What do you think of?

File: libvirt/tests/src/migration/migrate_network.py (hunk 1 above)
Comment (dzhengfy) [reply libvirt/tests/src/migration/migrate_network.py:28]: It sounds reasonable, too.

**Review by dzhengfy — COMMENTED** [review]
Yes. My understanding is that you want to remove all existing nics and add a new nic.
From my POV, modify_vm_iface() is messed up a little now. I am thinking if we need create a new function for creating a nic and reuse the assembling dict in modify_vm_iface(). What do you think of?

**Comment by Yingshun** [comment]
> Yes. My understanding is that you want to remove all existing nics and add a new nic.
> From my POV, modify_vm_iface() is messed up a little now. I am thinking if we need create a new function for creating a nic and reuse the assembling dict in modify_vm_iface(). What do you think of?

@dzhengfy I'm not sure what the 'mess up' mean is. Could you please explain more?

**Comment by dzhengfy** [comment]
> > Yes. My understanding is that you want to remove all existing nics and add a new nic.
> > From my POV, modify_vm_iface() is messed up a little now. I am thinking if we need create a new function for creating a nic and reuse the assembling dict in modify_vm_iface(). What do you think of?
> 
> @dzhengfy I'm not sure what the 'mess up' mean. Could you please explain more?

After discussion, we decided to keep as it is so far. We will file new PR to refine the function in better encapsulation.

**Comment by Yingshun** [comment]
> > > Yes. My understanding is that you want to remove all existing nics and add a new nic.
> > > From my POV, modify_vm_iface() is messed up a little now. I am thinking if we need create a new function for creating a nic and reuse the assembling dict in modify_vm_iface(). What do you think of?
> > 
> > 
> > @dzhengfy I'm not sure what the 'mess up' mean. Could you please explain more?
> 
> After discussion, we decided to keep as it is so far. We will file new PR to refine the function in better encapsulation.

@dzhengfy  could you please help review again?

**Review by dzhengfy — APPROVED** [review]
LGTM

**Comment by dzhengfy** [comment]
@kylazhang Could you help review? It is ovs related.

**Review by kylazhang — APPROVED** [review]

**Comment by kylazhang** [comment]
Approve and waiting for avocado-vt pr updated and merged first

**Review by kylazhang — APPROVED** [review]

**Pull request merged**